ipld/resolver/src/service.rs

```diff
@@ -1,17 +1,431 @@
+        match event {
+            content::Event::Complete(query_id, result) => {
+                if let Some(query) = self.queries.remove(&query_id) {
+                    self.handle_query_result(query, result);
+                } else {
+                    warn!("query ID not found");
+                }
+            }
+        }
+    }
+
+    /// Handle the results from a resolve attempt. If it succeeded, notify the
+    /// listener. Otherwise if we have fallback peers to try, start another
+    /// query and send the result to them. By default these are the peers
+    /// we know support the subnet, but weren't connected to when the we
+    /// first attempted the resolution.
+    fn handle_query_result(&mut self, mut query: Query, result: ResolveResult) {
+        match result {
+            Ok(_) => send_resolve_result(query.response_channel, result),
+            Err(_) if query.fallback_peer_ids.is_empty() => {
+                send_resolve_result(query.response_channel, result)
+            }
+            Err(e) => {
+                debug!(
+                    "resolving {} from {} failed with {}, but there are {} fallback peers to try",
+                    query.cid,
+                    query.subnet_id,
+                    e,
+                    query.fallback_peer_ids.len()
+                );
+
+                // Now we can go all in; alternatively we could take the next N peers.
+                let peers = std::mem::take(&mut query.fallback_peer_ids);
+
+                let query_id = self.content_mut().resolve(query.cid, peers);
+
+                self.queries.insert(query_id, query);
+            }
+        }
+    }
+
+    /// Handle an internal request coming from a [`Client`].
+    fn handle_request(&mut self, request: Request) {
+        match request {
+            Request::SetProvidedSubnets(ids) => {
+                if let Err(e) = self.membership_mut().set_provided_subnets(ids) {
+                    error!("error setting provided subnets: {e}")
+                }
+            }
+            Request::AddProvidedSubnet(id) => {
+                if let Err(e) = self.membership_mut().add_provided_subnet(id) {
+                    error!("error adding provided subnet: {e}")
+                }
+            }
+            Request::RemoveProvidedSubnet(id) => {
+                if let Err(e) = self.membership_mut().remove_provided_subnet(id) {
+                    error!("error removing provided subnet: {e}")
+                }
+            }
+            Request::PinSubnet(id) => self.membership_mut().pin_subnet(id),
+            Request::UnpinSubnet(id) => self.membership_mut().unpin_subnet(&id),
+
+            Request::Resolve(cid, subnet_id, response_channel) => {
+                let peers = self.membership_mut().providers_of_subnet(&subnet_id);
+                if peers.is_empty() {
+                    send_resolve_result(response_channel, Err(anyhow!(NoKnownPeers(subnet_id))));
```

Comment:
This would mean that no membership was picked up through gossipsub for the subnet, right? We could include in the future a fallback that optimistically asks our direct connections what subnets they are part of for the case we are lucky and we find a provider.

Reply:
Yes, most likely we asked too early and we should come back later, or perhaps find a known peer and add it to the static address list. 

I am also thinking on how to make the gossip better for newcomers, to lessen the trade-off of periodic republishing and let newcomers learn the providers ASAP.

Here's an idea: when we receive a gossip from a provider we haven't known about before, we publish our own in return. They must be a newcomer, so they could use our own information for sure. Perhaps with some cool-off period, like no repeated publishing within 10 seconds (just above the gossipsub broadcast time frame).

Reply:
Created https://github.com/consensus-shipyard/ipc-agent/issues/67